<?xml version="1.0"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro" name="magni">
  <xacro:property name="M_PI" value="3.1415926535897931" />
  
  <xacro:property name="wheel_r" value="0.1" />
  <xacro:property name="wheel_xpos" value="0.11" />

  <xacro:include filename="$(find magni_description)/urdf/magni.transmission.xacro" />
  <xacro:include filename="$(find magni_description)/urdf/magni.gazebo.xacro" />
  <xacro:include filename="$(find magni_description)/urdf/inertial.xacro" />

  <!-- Robot base_link -->
  <link name="base_link">
    <visual>
      <origin xyz="${-wheel_xpos} 0 ${-wheel_r+0.071}" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://magni_description/meshes/magni_body.dae" />
      </geometry>
    </visual>

    <!-- Base Chassis -->
    <collision>
	    <origin xyz="${-wheel_xpos} 0 ${-wheel_r+0.151}" rpy="0 0 0" />
      <geometry>
        <box size="0.398 0.268 0.160"/>
      </geometry>
    </collision>
     
    <!-- Top plate -->
    <collision>
      <origin xyz="${-wheel_xpos-0.026} 0 ${-wheel_r+0.248}" rpy="0 0 0" />
      <geometry>
        <box size="0.375 0.268 0.034"/>
      </geometry>
    </collision>
    <xacro:box_inertial_with_origin  x="0.398" y="0.268" z="0.160" mass="8">
      <origin xyz="-0.15 0 0.03" rpy="0 0 0" />
    </xacro:box_inertial_with_origin>
  </link>

  <!-- Fixed offset from the floor to the center of the wheels -->
  <link name="base_footprint" />
  <joint name="base_footprint_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link"/>
    <origin xyz="0 0 ${wheel_r}" rpy="0 0 0"/>
  </joint>

  <!-- Macro for defining wheels -->
  <xacro:macro name="wheel" params="prefix reflect">
    <link name="${prefix}_wheel">
      <visual>
        <origin xyz="${-0.071*reflect} 0 0" rpy="0 ${(reflect-1)*pi/2} 0" />
        <geometry>
          <mesh filename="package://magni_description/meshes/front_wheel.dae" />
        </geometry>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="0 ${pi/2} 0" />
        <geometry>
          <cylinder radius="${wheel_r}" length="0.05"/>
        </geometry>
      </collision>
      <collision>
        <origin xyz="${0.033*reflect} 0 0" rpy="0 ${pi/2} 0" />
        <geometry>
          <cylinder radius="0.064" length="0.016"/>
        </geometry>
      </collision>
      <xacro:cylinder_inertial_with_origin radius="${wheel_r}" length="0.05" mass="3.34">
        <origin xyz="0 0 0" rpy="0 ${pi/2} 0" />
      </xacro:cylinder_inertial_with_origin>
    </link>
    <joint name="${prefix}_wheel_joint" type="continuous">
      <axis xyz="1 0 0" rpy="0 0 0" />
      <parent link="base_link"/>
      <child link="${prefix}_wheel"/>
      <origin xyz="0 ${0.163*reflect} 0" rpy="0 0 ${pi/2}"/>
    </joint>
  </xacro:macro>

  <xacro:wheel prefix="left" reflect="1"/>
  <xacro:wheel prefix="right" reflect="-1"/>

  <!-- Macro for defining caster wheels -->
  <xacro:macro name="caster_wheel" params="prefix reflect">
    <link name="${prefix}_caster_wheel_mount">
      <collision>
        <!-- <origin xyz="${-wheel_xpos-0.169} ${0.169 * reflect} ${-wheel_r+0.106}" rpy="0 0 0" /> -->
        <geometry>
          <box size="0.1 0.07 0.03"/>
        </geometry>
      </collision>
    </link>

    <joint name="${prefix}_caster_mount_joint" type="fixed">
      <parent link="base_link"/>
      <child link="${prefix}_caster_wheel_mount"/>
      <origin xyz="${-wheel_xpos-0.169} ${0.169 * reflect} ${-wheel_r+0.106}" rpy="0 0 0" />
    </joint>

    <link name="${prefix}_caster_wheel">
      <visual>
        <!-- <origin xyz="${-wheel_xpos-0.169} ${reflect*0.169} ${-wheel_r+0.09}" rpy="0 0 ${M_PI}" /> -->
        <!-- <origin xyz="0.021 ${0.001 * reflect} 0.05" rpy="0 0 ${pi/2}" /> -->
        <!-- <origin xyz="${-0.01 * reflect} -0.026 0.053" rpy="0 0 ${pi/2}" /> -->
        <geometry>
          <mesh filename="package://magni_description/meshes/caster_wheel.dae" />
        </geometry>
      </visual> 
      <collision>
        <origin xyz="0.021 0 ${-wheel_r+0.0478}" rpy="0 0 0" />
        <geometry>
          <sphere radius="0.04"/>
        </geometry>
      </collision>
      <xacro:sphere_inertial_with_origin radius="0.04" mass="0.5" >
        <!-- <origin xyz="0.021 0 ${-wheel_r+0.0478}" rpy="0 0 0" /> -->
        <origin xyz="0 0 0" rpy="0 0 0" />
      </xacro:sphere_inertial_with_origin>
    </link>

    <joint name="${prefix}_caster_wheel_joint" type="fixed">
      <parent link="${prefix}_caster_wheel_mount"/>
      <child link="${prefix}_caster_wheel"/>
      <!-- <origin xyz="${-wheel_xpos-0.19} ${0.169 * reflect} ${-wheel_r+0.04}" rpy="0 0 ${pi/2}"/> -->
      <!-- <origin xyz="0.0 0.0 0.0" rpy="0 0 -${pi/2}"/> -->
      <origin xyz="0.0 0 -0.013" rpy="0 0 ${pi}" />
    </joint>
  </xacro:macro>

  <xacro:caster_wheel prefix="left" reflect="1"/>
  <xacro:caster_wheel prefix="right" reflect="-1"/>

  <!-- Velodyne support structure -->
  <link name="velodyne_support">
    <xacro:box_inertial_with_origin x="0.150028" y="0.150028" z="0.150028" mass="1e-9" >
      <origin xyz="0 0 0" rpy="0.0 0.0 0.0" />
    </xacro:box_inertial_with_origin>
    <visual>
      <origin xyz="0 0 0" rpy="0.0 0.0 0.0" />
      <geometry>
        <mesh filename="package://magni_octonav/meshes/support.stl" />
      </geometry>
      <material name="black" >
        <color rgba="0.2 0.2 0.2 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://magni_octonav/meshes/support.dae" />
      </geometry>
    </collision>
  </link>

  <!-- Velodyne URDF, plugin, etc -->
  <xacro:include filename="$(find magni_octonav)/urdf/VLP-16.urdf.xacro"/>
  <xacro:VLP-16 parent="velodyne_support" name="velodyne" topic="/velodyne_points" hz="10" gpu="true">
    <origin xyz="0 0 0.145028" rpy="0 0 0" />
  </xacro:VLP-16>

  <joint name="velodyne_support_joint" type="fixed">
    <parent link="base_link"/>
    <child link="velodyne_support"/>
    <origin xyz="-0.12 0 0.164" rpy="0 0 0"/>
  </joint>

</robot>